plugin/pkg/proxy/connect.go

```diff
@@ -121,9 +121,24 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 			// For UDP, if the error is not a network error keep waiting for a valid response to prevent malformed
 			// spoofs from blocking the upstream response.
 			// In the case this is a legitimate malformed response from the upstream, this will result in a timeout.
-			if proto == "udp" {
-				if _, ok := err.(net.Error); !ok {
-					continue
+
+			// If the error is an overflow, we probably have an upstream misbehaving in some way.
+			// (e.g. sending >512 byte UDP responses without an eDNS0 OPT RR).
+			// Instead of returning an error, return an empty response with TC bit set. This will make the
+			// client retry over TCP (if that's supported) or at least receive a clean
+			// error. The connection is still good so we break before the close.
+			if proto == "udp" && strings.Contains(err.Error(), "overflow") {
```

Comment:
Could we include `ErrBuf`? https://github.com/coredns/coredns/pull/6003/files#r1156290153

Reply:
This is done with below change - 

			dnsErrBufOccured := false
			var perr *dns.Error
			if errors.As(err, &perr) {
				if errors.Is(err, dns.ErrBuf) {
					dnsErrBufOccured = true
				}
			}

			if proto == "udp" && ((strings.Contains(err.Error(), "overflow")) || dnsErrBufOccured) {